pkg/operator/rules.go

```diff
@@ -108,27 +112,51 @@ func (prs *PrometheusRuleSelector) generateRulesConfiguration(promRule *monitori
 	return string(content), nil
 }
 
+// validateRuleFieldWithVersion checks whether fields added in Rule object is supported in respective Prometheus/Thanos versions.
+func (prs *PrometheusRuleSelector) validateRuleFieldsWithVersion(promRuleSpec monitoringv1.PrometheusRuleSpec) monitoringv1.PrometheusRuleSpec {
+	minVersionKeepFiringFor := semver.MustParse("2.42.0")
+
+	for i := range promRuleSpec.Groups {
+		for j := range promRuleSpec.Groups[i].Rules {
+			switch prs.ruleFormat {
+			case PrometheusFormat:
+				if promRuleSpec.Groups[i].Rules[j].KeepFiringFor != nil && prs.version.LT(minVersionKeepFiringFor) {
+					promRuleSpec.Groups[i].Rules[j].KeepFiringFor = nil
+					level.Warn(prs.logger).Log("msg", fmt.Sprintf("`keep_firing_for` is supported only from %s version %q", "prometheus", minVersionKeepFiringFor))
+				}
+			case ThanosFormat:
+				// keep_firing_for is not yet supported in thanos https://github.com/thanos-io/thanos/issues/6165
+				if promRuleSpec.Groups[i].Rules[j].KeepFiringFor != nil {
+					promRuleSpec.Groups[i].Rules[j].KeepFiringFor = nil
+					level.Warn(prs.logger).Log("msg", "`keep_firing_for` is not yet supported in thanos check https://github.com/thanos-io/thanos/issues/6165")
+				}
+			}
+		}
+	}
+	return promRuleSpec
+}
+
 // ValidateRule takes PrometheusRuleSpec and validates it using the upstream prometheus rule validator.
-func ValidateRule(promRule monitoringv1.PrometheusRuleSpec) []error {
-	for i := range promRule.Groups {
+func ValidateRule(promRuleSpec monitoringv1.PrometheusRuleSpec) []error {
```

Comment:
I didnt use this func because its used by validating webhook as well which wont pass prometheus version. Also dont think we need version checks in webhook

Reply:
correct